#163 Refactor lineup assistant

wsbrenk opened the pull request:
1. changes proposed in this pull request:

   - fixes issue #163
   - if not fixing an existing issue, comments explaining the purpose of the PR should be provided)


2. `src/main/resources/release_notes.md` ...

 - [x] has been updated
 - [ ] does not require update

tychobrailleur commented on src/main/java/module/lineup/assistant/LineupAssistant.java:
```diff
@@ -53,159 +53,250 @@ public final boolean isPlayerInStartingEleven(int spielerId, Vector<MatchLineupP
 
-		lPositions = filterPositions(lPositions);
+		var lPositions = allPositions.stream().filter(i -> i.isFieldMatchRoleId() && isSelectedByAssistant(i)).toList();
+		ArrayList<Player> players = new ArrayList<>(lPlayers.stream().filter(i -> (i.getInjuryWeeks() < 1 || ignoreInjured) && (!i.isRedCarded() || ignoreSuspended)).toList());
```
```
List<Player> players = lPlayers.stream()  ... .toList();
```
should be enough (i.e. without additional `new ArrayList<>`) – then you just need to change the signature of `optimizeLineup` to take a `List<Player> players`.

tychobrailleur replied:
(or replace `toList()` with `.collect(Collectors.toList())` for a mutable list)

wsbrenk replied:
mutable list is required - i will try your 2nd suggestion - Thanks!

tychobrailleur commented on src/main/java/module/lineup/assistant/LineupAssistant.java:
```diff
@@ -53,159 +53,250 @@ public final boolean isPlayerInStartingEleven(int spielerId, Vector<MatchLineupP
+				var maxRating = -1.0;
+				Player bestPlayer = null;
+				byte bestBehaviour = 0;
+				for (var behaviour : behaviours) {
```
Wow!!  So this computes the best attitude for the player in the position?  That's pretty cool!

wsbrenk replied:
Attitude or Behaviour? What would be the better name for it?

tychobrailleur replied:
> Attitude or Behaviour? What would be the better name for it?

Hattrick calls them “individual orders.” Behaviour is fine!

tychobrailleur commented on src/main/java/module/lineup/assistant/LineupAssistant.java:
```diff
@@ -53,159 +53,250 @@ public final boolean isPlayerInStartingEleven(int spielerId, Vector<MatchLineupP
+					pos.setPlayerIdIfValidForLineup(bestPlayer.getPlayerId());
+					pos.setBehaviour(bestBehaviour);
+					if (!pos.isBackupsMatchRoleId()) {
+						players.remove(bestPlayer);
```
One thing that troubles me, and you're probably better placed here to comment, is that we seem to give precedence to the order in which we find players in the list, which may not be the optimal one (another player further down the list could  be better in an earlier position, and the resulting line-up could overall be better).  Is this a a fair thing to say?  

Obviously exploring multiple orders would be expensive from a resource perspective, but I wonder if that could be optimised somehow.

tychobrailleur replied:
Another thing I notice is that the attitude selection maximises the player rating at a given position, but doesn't necessarily maximise the expected sector: I have an example of a winger that has a higher rating as defensive, but when set to normal increases midfield slightly, and decreases defence a fair bit.  This makes total sense, but I wonder this is a behaviour the user would expect?

There could be a separate option in the future that aims to maximise selected sector (maybe based on the choice AW/MF/ST?), but probably not relevant at this point.

wsbrenk replied:
The order of the player list has no impact to the optimized lineup - only the order of the positions/sectors.

The calculated player rating is quite similar to the hatstat calculation. I fear it would be very hard to find further improvements. With the help of the sector preference setting, the desired sector is already preferred.

tychobrailleur commented on src/main/java/module/lineup/assistant/LineupAssistant.java:
```diff
@@ -53,159 +53,250 @@ public final boolean isPlayerInStartingEleven(int spielerId, Vector<MatchLineupP
+		var reservePositionOrder = new ArrayList<List<Byte>>();
+		fieldPlayerPositionOrder.add(List.of(IMatchRoleID.KEEPER));
+		reservePositionOrder.add(List.of(IMatchRoleID.KEEPER));
 		switch (sectorsStrengthPriority) {
```
Do you think it'd be possible some day to compute all 6 combinations in parallel, and identify the best approach for all 3 sectors?  Or better even let the user select _a posteriori_ what combination to pick for whatever sector they want maximised.

That would have a significant impact on computation, but also on UI, I think.

wsbrenk replied:
Add "Auto" Entry to the selection list  - good idea

tychobrailleur commented on src/main/java/module/lineup/assistant/LineupAssistant.java:
```diff
@@ -53,159 +53,250 @@ public final boolean isPlayerInStartingEleven(int spielerId, Vector<MatchLineupP
-	public final void doLineup(List<MatchLineupPosition> lPositions, List<Player> lPlayers,
-							   byte sectorsStrengthPriority, boolean bForm, boolean idealPosFirst, boolean bInjured,
-							   boolean bSuspended, Weather weather) {
+	public final void doLineup(List<MatchLineupPosition> allPositions, List<Player> lPlayers,
```
Here I am thinking of working myself on a unit test framework that would allow us to test these things.

wsbrenk replied:
Arguments form and weather are currently not used.

tychobrailleur commented on src/main/resources/release_notes.md:
```diff
@@ -35,6 +35,7 @@
 
 ### Lineup
 
+* Refactoring of lineup assistant (#163)
```
I think it's more than refactoring the assistant, it's adding an additional feature, right?

wsbrenk replied:
Changed this.

tychobrailleur reviewed (approved):
Thanks a lot for this change, really cool improvement!

tychobrailleur commented:
Following this change, I see locally some of the substitutes that are benched are not only players that are in the starting 11, but also placed in multiple positions.
![Screenshot From 2025-01-07 09-41-41](https://github.com/user-attachments/assets/e0fa33de-b6dd-4c1a-834d-1445debc990b)

wsbrenk commented:
@tychobrailleur can not reproduce this - can it be, that you are working with a not mutable player list?

List<Player> players = lPlayers.stream()  ... .toList();

tychobrailleur commented:
> @tychobrailleur can not reproduce this - can it be, that you are working with a not mutable player list?
> 
> List players = lPlayers.stream() ... .toList();

I don't believe I made any change to the code, but I'll check later on.

wsbrenk commented:
now i updated my productive version and saw a similar error - although i only saw duplicates on the backup positions, the lineup was not accepted by hattrick, telling about duplicate player positions - even if i fixed it by clearing the backup positions.

tychobrailleur commented:
> now i updated my productive version and saw a similar error - although i only saw duplicates on the backup positions, the lineup was not accepted by hattrick, telling about duplicate player positions - even if i fixed it by clearing the backup positions.

I cleaned and re-compiled, and the issue went away -- apologies for the noise, it must have been some cached classes from my experiments.
I do also see duplicates in the backup line, but Hattrick accepts mine:

![Screenshot From 2025-01-08 20-19-39](https://github.com/user-attachments/assets/cb9d208c-f48a-40c0-9f72-ab22bfaa6c2b)

wsbrenk commented:
yes - today my upload was fine too - maybe a hattrick glitch
